src/Facets/GenericSwapFacetV3.sol

```diff
@@ -0,0 +1,515 @@
+            revert CumulativeSlippageTooHigh(_minAmountOut, amountReceived);
+
+        // transfer funds to receiver
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success, ) = _receiver.call{ value: amountReceived }("");
+        if (!success) revert NativeAssetTransferFailed();
+
+        // emit event
+        emit LiFiGenericSwapCompleted(
+            _transactionId,
+            _integrator,
+            _referrer,
+            _receiver,
+            _swapData[0].sendingAssetId,
+            address(0),
+            _swapData[0].fromAmount,
+            amountReceived
+        );
+    }
+
+    function _depositAndSwapERC20Single(
+        LibSwap.SwapData calldata _swapData,
+        address _receiver
+    ) private {
+        ERC20 sendingAsset = ERC20(_swapData.sendingAssetId);
+        uint256 fromAmount = _swapData.fromAmount;
+        // deposit funds
+        sendingAsset.safeTransferFrom(msg.sender, address(this), fromAmount);
+
+        // ensure that contract (callTo) and function selector are whitelisted
+        address callTo = _swapData.callTo;
+        address approveTo = _swapData.approveTo;
+        bytes calldata callData = _swapData.callData;
+        if (
+            !(LibAllowList.contractIsAllowed(callTo) &&
+                LibAllowList.selectorIsAllowed(bytes4(callData[:4])))
+        ) revert ContractCallNotAllowed();
+
+        // ensure that approveTo address is also whitelisted if it differs from callTo
+        if (approveTo != callTo && !LibAllowList.contractIsAllowed(approveTo))
+            revert ContractCallNotAllowed();
+
+        // check if the current allowance is sufficient
+        uint256 currentAllowance = sendingAsset.allowance(
+            address(this),
+            approveTo
+        );
+
+        // check if existing allowance is sufficient
+        if (currentAllowance < fromAmount) {
+            // check if is non-zero, set to 0 if not
+            if (currentAllowance != 0) sendingAsset.safeApprove(approveTo, 0);
+            // set allowance to uint max to avoid future approvals
+            sendingAsset.safeApprove(approveTo, type(uint256).max);
+        }
+
+        // execute swap
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success, bytes memory res) = callTo.call(callData);
+        if (!success) {
+            string memory reason = LibUtil.getRevertMsg(res);
+            revert(reason);
+        }
+
+        _returnPositiveSlippage(sendingAsset, _receiver);
+    }
+
+    function _returnPositiveSlippage(
+        ERC20 fromToken,
+        address receiver
+    ) private {
+        // if a balance exists in fromToken, it must be positive slippage
+        if (address(fromToken) != address(0)) {
+            uint256 fromTokenBalance = fromToken.balanceOf(address(this));
+            if (fromTokenBalance > 0) {
+                fromToken.safeTransfer(receiver, fromTokenBalance);
+            }
```

Comment:
Why does it not handle native tokens?

Reply:
@maxklenk: 
Because we send native as msg.value in full to DEXs.....are you aware of any cases where a DEX would not use that msg.value in full but send a part back as a unused refund? If we have such a case, then it would make sense to add handling for it. I just wasnt aware of any such cases.
Please let me know.

Reply:
If we want to use Uniswap V3 in the future we will need to be able to handle native token refunds. 0x actually facilitates UniV3 swaps for us sometimes which result in refunds of native token.

Reply:
The only case where refunds can happen is with uniswapV3, and as Ed said I would expect it for all tokens. I think it does not hurt to have a fallback for the native case here as well.

Reply:
I've added handling for native sendingAsset refunds by a DEX